File: 6.py
def getSubString(string:str,index:int)->str:
    if(index==0):
        return str[1:]
    if(index==len(str)-1):
        return str[:-1]
    return str[:index] + str[index+1:]
def insertChar2String(string:str,insert:str,index:int)->str:
    if(index==0):
        return insert+string
    if(index == len(string)):
        return string + insert
    return string[:index] + insert + string[index:]
def isLegal(string:str)->bool:
    n = 0
    for char in string:
        if(char == "("):
            n+=1
        elif(char == ")"):
            n-=1
            if(n<0):return False
    return n==0

if __name__ == '__main__':
    str = input()
    # str = "(())"
    dict = {}
    for i in range(len(str)):
        temp = getSubString(str,i)
        for j in range(len(temp)+1):
            insertString = insertChar2String(temp,str[i],j)
            if(insertString!=str and isLegal(insertString)):
                dict[insertString] = 0
    print(len(dict))

File: 6.cpp
//
// Created by zb198 on 2019/8/28.
//
#include <iostream>
#include <string>
#include <map>
bool isLegal(std::string str){
    int n=0;
    int length = str.size();
    for(int i=0;i<length;i++){
        if(str[i]=='('){
            n++;
        } else if(str[i]==')'){
            n--;
            if(n<0){
                return false;
            }
        }
    }
    return n==0;
}
std::string getSubString(std::string* str,int index){
    return str->substr(0,index) + str->substr(index+1,str->size()-index-1);
}
std::string insertString(std::string* str,int index,char insert){
    return str->substr(0,index) + insert + str->substr(index,str->size()-index);
}
int main(){
    std::map<std::string,int> hash;
    std::string str;
    std::cin>>str;
//    str = "(())";
    int n = str.size();
    for(int i=0;i<n;i++){
        std::string temp = getSubString(&str,i);
        for(int j=0;j<n-1;j++) {
            char stri = str[i];
            std::string insert = insertString(&temp,j,stri);
            if (insert != str && isLegal(insert)) {
                hash[insert] = 0;
            }
        }
    }
    std::cout<<hash.size();
    return 0;
}
